python/ray/util/client/server/proxier.py

```diff
@@ -0,0 +1,323 @@
+from concurrent import futures
+from dataclasses import dataclass
+import grpc
+import logging
+import json
+from queue import Queue
+import random
+import socket
+from threading import Thread, Lock
+import time
+from typing import Any, Callable, Dict, Iterator, Optional, Tuple
+
+from ray.job_config import JobConfig
+import ray.core.generated.ray_client_pb2 as ray_client_pb2
+import ray.core.generated.ray_client_pb2_grpc as ray_client_pb2_grpc
+from ray.util.client.common import (ClientServerHandle,
+                                    CLIENT_SERVER_MAX_THREADS, GRPC_OPTIONS)
+from ray._private.services import ProcessInfo, start_ray_client_server
+
+logger = logging.getLogger(__name__)
+
+CHECK_PROCESS_INTERVAL_S = 30
+
+MIN_SPECIFIC_SERVER_PORT = 23000
+MAX_SPECIFIC_SERVER_PORT = 24000
+
+CHECK_CHANNEL_TIMEOUT_S = 5
+
+
+def _get_unused_port() -> int:
```

Comment:
You are probably aware that this introduces a race condition?

Reply:
One way to fix this would be to use unix domain sockets of the form `unix:///tmp/ray/sockets/{uuid}` if Python GRPC supports them.

Reply:
@pcmoritz Yeah. This could definitely be a danger if the server takes a while to start up (because startup might require installing a conda environment).

> Unix Domains

Python gRPC **does** support them, but I would prefer to stick with network ports because it allows for direct connections.

I am leaning towards having an internal list of ports in the ProxyManager. What are your thoughts here?

Reply:
Ah dang about the unix domain sockets.

A range of ports like in the core worker (and non-overlapping with the core worker ones) sounds great to me! Not exactly sure how we allocate ports there, but taking consecutive ports in that range and skipping the ones that are taken makes sense :)

Reply:
Yep! Will do something like that logic (i.e. have a free list)